Source/coreutils/factor/factor.1
.\"	$NetBSD: factor.6,v 1.18 2020/10/12 13:53:43 christos Exp $
.\"
.\" Copyright (c) 1989, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" This code is derived from software contributed to Berkeley by
.\" Landon Curt Noll.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	@(#)factor.6	8.1 (Berkeley) 5/31/93
.\"
.\"
.\" By Landon Curt Noll, http://www.isthe.com/chongo/index.html /\oo/\
.\"
.Dd October 12, 2020
.Dt FACTOR 1
.Os
.Sh NAME
.Nm factor
.Nd factor a number
.Sh SYNOPSIS
.Nm
.Op Fl hx
.Op Ar number ...
.Sh DESCRIPTION
The
.Nm
utility factors integers larger than 
.Dv 0 .
When a number is factored, it is printed, followed by a
.Dq \&: ,
and the list of
.Pq prime
factors on a single line.
Factors are listed in ascending order, and are preceded by a space.
By default, if a factor divides a value more than once, it will be
printed more than once.
.Pp
When
.Nm
is invoked with one or more
.Ar number
arguments, each argument will be factored.
.Pp
When
.Nm
is invoked with no
.Ar number
arguments,
.Nm
reads numbers, one per line, from standard input, until end of file or error.
Leading white-space and empty lines are ignored.
Numbers may be preceded by a single +.
Numbers that start with
.Dv 0x
are interpreted as hexadecimal.
A 
.Dv 0
input exits the program.
Negative numbers are not allowed.
Numbers are terminated by a non-digit character (such as a newline).
After a number is read, it is factored.
Input lines must not be longer than
.Dv LINE_MAX \- 1
(currently 2047) characters.
.Pp
By default,
.Nm
is compiled against the OpenSSL bignum implementation
.Xr openssl_bn 3 ,
which lets it handle arbitrarily large values.
.Pq Note however that very large values can take a very long time to factor.
If
.Nm
is compiled without OpenSSL it is limited to the maximum value of
.Vt unsigned long .
.Pp
The following option is available:
.Bl -tag -width flag
.It Fl h
If the
.Fl h
flag is specified, factors will be printed in "human-readable" format.
If a factor x divides a value n (>1) times, it will appear as x^n.
.It Fl x
If the
.Fl x
flag is specified, factors will be printed in hexadecimal format.
.El
.Sh DIAGNOSTICS
Out of range or invalid input results in
an appropriate error message to standard error.
.Sh AUTHORS
Originally by
.An Landon Curt Noll .


Source/coreutils/factor/factor.c
/*	$NetBSD: factor.c,v 1.38 2020/10/12 13:54:51 christos Exp $	*/
/*
 * Copyright (c) 1989, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * This code is derived from software contributed to Berkeley by
 * Landon Curt Noll.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

/*
 * factor - factor a number into primes
 *
 * By: Landon Curt Noll   chongo@toad.com,   ...!{sun,tolsoft}!hoptoad!chongo
 *
 *   chongo <for a good prime call: 391581 * 2^216193 - 1> /\oo/\
 *
 * usage:
 *	factor [-hx] [number] ...
 *
 * The form of the output is:
 *
 *	number: factor1 factor1 factor2 factor3 factor3 factor3 ...
 *
 * where factor1 <= factor2 <= factor3 <= ...
 *
 * If the -h flag is specified, the output is in "human-readable" format.
 * Duplicate factors are printed in the form of x^n.
 *
 * If the -x flag is specified numbers are printed in hex.
 *
 * If no number args are given, the list of numbers are read from stdin.
 */

#include <ctype.h>
#include <err.h>
#include <errno.h>
#include <inttypes.h>
#include <limits.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

#include "primes.h"

#include <openssl/bn.h>

#define	PRIME_CHECKS	5

/* print factors for big numbers */
static void	pollard_pminus1(BIGNUM *, int, int);

static void	BN_print_dec_fp(FILE *, const BIGNUM *);
static void	convert_str2bn(BIGNUM **, char *);
static void	pr_fact(BIGNUM *, int, int);	/* print factors of a value */
static void	pr_print(BIGNUM *, int, int);	/* print a prime */
static void	usage(void) __attribute__((noreturn));

static BN_CTX	*ctx;			/* just use a global context */

int
main(int argc, char *argv[])
{
	BIGNUM *val;
	int ch, hflag, xflag;
	char *p, buf[LINE_MAX];		/* > max number of digits. */

	hflag = xflag = 0;
	ctx = BN_CTX_new();
	val = BN_new();
	if (val == NULL)
		errx(1, "can't initialise bignum");

	while ((ch = getopt(argc, argv, "hx")) != -1)
		switch (ch) {
		case 'h':
			hflag++;
			break;
		case 'x':
			xflag++;
			break;
		case '?':
		default:
			usage();
		}
	argc -= optind;
	argv += optind;

	/* No args supplied, read numbers from stdin. */
	if (argc == 0)
		for (;;) {
			if (fgets(buf, sizeof(buf), stdin) == NULL) {
				if (ferror(stdin))
					err(1, "stdin");
				exit (0);
			}
			for (p = buf; isblank((unsigned char)*p); ++p);
			if (*p == '\n' || *p == '\0')
				continue;
			convert_str2bn(&val, p);
			pr_fact(val, hflag, xflag);
		}
	/* Factor the arguments. */
	else
		for (p = *argv; p != NULL; p = *++argv) {
			convert_str2bn(&val, p);
			pr_fact(val, hflag, xflag);
		}
	exit(0);
}

static void
pr_exp(int i, int xflag)
{
	printf(xflag && i > 9 ? "^0x%x" : "^%d", i);
}

static void
pr_uint64(uint64_t i, int xflag)
{
	printf(xflag ? " 0x%" PRIx64 : " %" PRIu64, i);
}

/*
 * pr_fact - print the factors of a number
 *
 * Print the factors of the number, from the lowest to the highest.
 * A factor will be printed multiple times if it divides the value
 * multiple times.
 *
 * If hflag is specified, duplicate factors are printed in "human-
 * readable" form of x^n.
 *
 * If the xflag is specified, numbers will be printed in hex.
 *
 * Factors are printed with leading tabs.
 */
static void
pr_fact(BIGNUM *val, int hflag, int xflag)
{
	int i;
	const uint64_t *fact;	/* The factor found. */

	/* Firewall - catch 0 and 1. */
	if (BN_is_zero(val))	/* Historical practice; 0 just exits. */
		exit(0);
	if (BN_is_one(val)) {
		printf("1: 1\n");
		return;
	}

	/* Factor value. */

	if (xflag) {
		fputs("0x", stdout);
		BN_print_fp(stdout, val);
	} else
		BN_print_dec_fp(stdout, val);
	putchar(':');
	fflush(stdout);
	for (fact = &prime[0]; !BN_is_one(val); ++fact) {
		/* Look for the smallest factor. */
		do {
			if (BN_mod_word(val, (BN_ULONG)*fact) == 0)
				break;
		} while (++fact <= pr_limit);

		/* Watch for primes larger than the table. */
		if (fact > pr_limit) {
			BIGNUM *bnfact;

			bnfact = BN_new();
			BN_set_word(bnfact, *(fact - 1));
			if (!BN_sqr(bnfact, bnfact, ctx))
				errx(1, "error in BN_sqr()");
			if (BN_cmp(bnfact, val) > 0 ||
			    BN_is_prime_ex(val, PRIME_CHECKS, NULL, NULL) == 1)
				pr_print(val, hflag, xflag);
			else
				pollard_pminus1(val, hflag, xflag);
			pr_print(NULL, hflag, xflag);
			break;
		}

		/* Divide factor out until none are left. */
		i = 0;
		do {
			i++;
			if (!hflag)
				pr_uint64(*fact, xflag);
			BN_div_word(val, (BN_ULONG)*fact);
		} while (BN_mod_word(val, (BN_ULONG)*fact) == 0);

		if (hflag) {
			pr_uint64(*fact, xflag);
			if (i > 1)
				pr_exp(i, xflag);
		}

		/* Let the user know we're doing something. */
		fflush(stdout);
	}
	putchar('\n');
}

static void
pr_print(BIGNUM *val, int hflag, int xflag)
{
	static BIGNUM *sval;
	static int ex = 1;
	BIGNUM *pval;

	if (hflag) {
		if (sval == NULL) {
			sval = BN_dup(val);
			return;
		}

		if (val != NULL && BN_cmp(val, sval) == 0) {
			ex++;
			return;
		}
		pval = sval;
	} else if (val == NULL) {
		return;
	} else {
		pval = val;
	}

	if (xflag) {
		fputs(" 0x", stdout);
		BN_print_fp(stdout, pval);
	} else {
		putchar(' ');
		BN_print_dec_fp(stdout, pval);
	}

	if (hflag) {
		if (ex > 1)
			pr_exp(ex, xflag);

		if (val != NULL) {
			BN_copy(sval, val);
		} else {
			BN_free(sval);
			sval = NULL;
		}
		ex = 1;
	}
}

static void
usage(void)
{
	fprintf(stderr, "Usage: %s [-hx] [value ...]\n", getprogname());
	exit(1);
}

/* pollard p-1, algorithm from Jim Gillogly, May 2000 */
static void
pollard_pminus1(BIGNUM *val, int hflag, int xflag)
{
	BIGNUM *base, *rbase, *num, *i, *x;

	base = BN_new();
	rbase = BN_new();
	num = BN_new();
	i = BN_new();
	x = BN_new();

	BN_set_word(rbase, 1);
newbase:
	if (!BN_add_word(rbase, 1))
		errx(1, "error in BN_add_word()");
	BN_set_word(i, 2);
	BN_copy(base, rbase);

	for (;;) {
		BN_mod_exp(base, base, i, val, ctx);
		if (BN_is_one(base))
			goto newbase;

		BN_copy(x, base);
		BN_sub_word(x, 1);
		if (!BN_gcd(x, x, val, ctx))
			errx(1, "error in BN_gcd()");

		if (!BN_is_one(x)) {
			if (BN_is_prime_ex(x, PRIME_CHECKS, NULL, NULL) == 1)
				pr_print(x, hflag, xflag);
			else
				pollard_pminus1(x, hflag, xflag);
			fflush(stdout);

			BN_div(num, NULL, val, x, ctx);
			if (BN_is_one(num))
				return;
			if (BN_is_prime_ex(num, PRIME_CHECKS, NULL,
			    NULL) == 1) {
				pr_print(num, hflag, xflag);
				fflush(stdout);
				return;
			}
			BN_copy(val, num);
		}
		if (!BN_add_word(i, 1))
			errx(1, "error in BN_add_word()");
	}
}

/*
 * Sigh..  No _decimal_ output to file functions in BN.
 */
static void
BN_print_dec_fp(FILE *fp, const BIGNUM *num)
{
	char *buf;

	buf = BN_bn2dec(num);
	if (buf == NULL)
		return;	/* XXX do anything here? */
	fprintf(fp, "%s", buf);
	free(buf);
}

/* Convert string pointed to by *str to a bignum.  */
static void
convert_str2bn(BIGNUM **val, char *p)
{
	int n = 0;

	if (*p == '+') p++;
	if (*p == '-')
		errx(1, "negative numbers aren't permitted.");
	if (*p == '0' && (p[1] == 'x' || p[1] == 'X')) {
		n = BN_hex2bn(val, p + 2);
	} else {
		n = BN_dec2bn(val, p);
	}
	if (n == 0)
		errx(1, "%s: illegal numeric format.", p);
}
